common/flatpak-transaction.c

```diff
@@ -2879,6 +2900,89 @@ flatpak_transaction_run (FlatpakTransaction *transaction,
   return FLATPAK_TRANSACTION_GET_CLASS (transaction)->run (transaction, cancellable, error);
 }
 
+static gboolean
+get_installation_dir_free_space (FlatpakInstallation  *installation,
+                                 guint64              *free_space,
+                                 GError              **error)
+{
+  g_autoptr (GFile) installation_dir = NULL;
+  g_autoptr (GFileInfo) info = NULL;
+
+  installation_dir = flatpak_installation_get_path (installation);
+
+  info = g_file_query_filesystem_info (installation_dir,
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_FREE,
+                                       NULL,
+                                       error);
+  if (info == NULL)
+    return FALSE;
+
+  *free_space = g_file_info_get_attribute_uint64 (info, G_FILE_ATTRIBUTE_FILESYSTEM_FREE);
+  return TRUE;
+}
+
+static gboolean
+check_for_disk_space (FlatpakTransaction *self)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_autoptr (GFile) installation_dir = NULL;
+  g_autoptr (GError) error = NULL;
+  GList *l;
+  guint64 space_required = 0;
+  guint64 min_free_space = 0;
+  guint64 free_disk_space = 0;
+
+  g_debug ("Checking for free disk space for the current transaction");
+
+  for (l = priv->ops; l != NULL; l = l->next)
+    {
+      FlatpakTransactionOperation *op = l->data;
+
+      if (op->kind == FLATPAK_TRANSACTION_OPERATION_INSTALL)
+        space_required += flatpak_transaction_operation_get_installed_size (op);
+      else if (op->kind == FLATPAK_TRANSACTION_OPERATION_UPDATE)
+        {
+          /* Runtime updates typically are small hence keep their sizes out of
```

Comment:
This isn't actually true on Endless. `com.endlessm.encyclopedia.en.Content/x86_64/eos3` is a runtime, and it's 5.6 GB. One might expect it to have large-ish updates.

Reply:
I know. And I don't have a good case to defend here. This is what we used to follow in our downstream [heuristics](https://github.com/endlessm/gnome-software/commit/5406424a) in gnome-software.

Reply:
In this case, the GsApp representing the app is always installed/updated along with its related refs such as the extensions, so the desired behaviour was achieved (that we would be more pessimistic about app updates, and more optimistic about runtimes). To achieve the same heuristic inside FlatpakTransaction (deciding our policy for space requirements depending if a runtime was added as a related ref, or as the thing we're adding directly) then we would probably need to be assembling a counter of the required space as we were adding the related refs to the transaction... or something?!

Reply:
What do you mean by assembling a counter "of the required space" ? I didn't get that properly.

Also, what would that counter do exactly for computing the space required?

Reply:
> What do you mean by assembling a counter "of the required space" ? I didn't get that properly.

Inside (our old) GNOME Software, when we consider upgrading a GsApp that represents an app, we know when we're looking at its related refs (extensions) that the end goal is to install an app - so we can choose a different disk space heuristic to when we're installing a GsApp that reflects a runtime. Inside FlatpakTransaction, because the operations are decomposed earlier - ie, if you add a app to the transaction, many runtime update or install ops are also added for the related refs, you can't tell by looking at them (one by one) whether it was added as the related ref of an app, or because the user is trying to explicitly act on a runtime.

> Also, what would that counter do exactly for computing the space required?

The same computation as you have implemented here, but done incrementally - by keeping some state (eg, required_disk) in the FlatpakTransaction which is a cumulative total of the amount of space that must be free for the tralsaction to continue. When we add the related refs for an app, we would increase this total, but when we added a runtime directly, we would not. So we'd avoid the fact that by the time we're running the transaction, we don't know any more if a "runtime" was added as a related ref, or because we want to upgrade a runtime directly.

Reply:
Well, we could easily track in FlatpakTransaction which add()s were from dependencies and which were from the API.